Move the full contribution guide for new 3rd party app support to GitHub. (#158)
GoogleCloudPlatform/ops-agent: dev-docs/new-receivers.md

# New 3rd Party Application Support How-To

## Steps

### 1. Research what telemetry the application generates

Figure out the following details for the application to facilitate the config design later.

> - How many versions of the application are dominating in the industry, and whether we need to support more than one versions.

Reviewer: ```suggestion
- How many versions of the application are dominating in the industry, and whether we need to support more than one version.
```

- For each of the version we need to support, figure out:
  - Logging
    - What logs: Once installed, what type of logs the application writes by default (e.g. Apache access logs and Apache
      error logs). Any additional application settings users can change to enable additional logging.
    - Log file paths: The common log file paths for these logs.
    - Log formats: The log formats inside these files, and available customization options if users want to change the
      log format.
  - Metrics
    - What metrics: Once installed, what metrics the application exposes by default. Any additional application settings
      users can change to expose additional metrics.
    - How to expose metrics. If the application does not expose metrics by default (e.g. Apache does not), what
      application settings users need to specify to enable it.
    - Required user configurations (e.g. username) for the Ops Agent to talk to the application to get the metrics

>       (e.g. username, password, database name, port etc.)

Reviewer: ```suggestion
      (e.g. username, password, database name, port, etc.)
```

  - If there are required user settings (e.g. username, password) for the Ops Agent to talk to the application, the metrics receiver(s) should have corresponding parameter.
  - If we need to support more than one versions of the application, and the metrics setup vary significantly across versions, the metrics receivers should have a corresponding `version` parameter. And the way this receiver works might change based on the value of the `version` parameter.

Take Apache for example

Given the research result above, there will be 1 new metrics receiver (type: `apache`) and 2 logging receivers (type:
`apache_access` and `apache_error`) for Apache. A sample config snippet (detailed parameters are TBD) looks like:

```
logging:
  receivers:
    apache_access:
      type: apache_access
      included_paths: /var/log/my_access_log # Optional.
      format: # Optional
    apache_error:
      type: apache_error
      included_paths: /var/log/my_error_log # Optional.
      format: # Optional
  service:
    pipelines:
      apache:

>         receivers:
>         - apache_access
>         - apache_error

Reviewer: ```suggestion
        receivers: [apache_access, apache_error]
```

metrics:
  receivers:
    apache:
      type: apache
      endpoint: #  Optional
      username: # Optional
      password: # Optional
  service:
    pipelines:
      apache:

>         receivers:
>         - apache

Reviewer: ```suggestion
        receivers: [apache]
```

```

#### 2.3 Validations against invalid configurations for the receiver(s)

Sample validation cases are like:

> - A required parameter is not set for the given receiver (s)

Reviewer: ```suggestion
- A required parameter is not set for the given receiver(s)
```

> - An unknown parameter is set for the receiver (s)

Reviewer: ```suggestion
- An unknown parameter is set for the receiver(s)
```

Follow https://github.com/GoogleCloudPlatform/ops-agent/blob/master/integration_test/README.md for detailed instructions.

### 6. Release and Public documentation

To prepare for a formal release that includes support for this application, the corresponding public documentation needs
to be ready:

- Public documentation: https://cloud.google.com/stackdriver/docs/solutions/agents/ops-agent.
- Configuration specific page: https://cloud.google.com/stackdriver/docs/solutions/agents/ops-agent/configuration.

TODO: Create a designated page / section for 3rd party applications in the cloud.google.com site above and link it to
application specific instructions inside this GitHub repo.

For each application, the documentation needs to cover the following:
- How to enable logs / metrics ingestion for this application
  - If the application does not expose metrics by default, what users need to do to the application to enable it. This
    might be no-op for many applications.
  - How users customize the Ops Agent config file to enable scraping and ingesting telemetry for this application. This
    contains details about the new receiver type(s). For example:

>   - The default log file paths the logging receiver(s) expect logs to be at
>   - The default log formats the logging receiver(s) expect the logs to be in
>   - The default metrics endpoint / port the metrics receiver expect metrics to be exposed at
>   - Available parameters of the receiver(s) to control additional behaviors

Reviewer: ```suggestion
    - The default log file paths the logging receiver(s) expect logs to be at
    - The default log formats the logging receiver(s) expect the logs to be in
    - The default metrics endpoint / port the metrics receiver expect metrics to be exposed at
    - Available parameters of the receiver(s) to control additional behaviors
```

  - Sample pipeline and receiver configuration snippets
- How to query logs and metrics for this application once ingested
  - When it comes to querying the logs and metrics data points, which log name, log labels, metrics namespace, metrics
    names, metrics labels to filter by.
  - Which dashboards contain these metrics by default.

> - Common failure cases specific to this application and how to troubleshoot them

Reviewer: ```suggestion
- Common failure cases specific to this application and how to troubleshoot them.
```

  For each of the failure cases, document how users could detect that failure condition (e.g. via errors in the Ops

>   Agent log, via Ops Agent’s own health metrics), and how to fix it. Sample failure condition includes:

Reviewer: ```suggestion
  Agent log, via Ops Agent’s own health metrics), and how to fix it. Sample failure conditions include:
```

- One set of code to generate otel configs from a config struct
- One set of code to generate fluentbit configs from a config struct

And then one file per 3rd party application integration should contain the config struct(s) for that integration and any
integration-specific business logic. These files are defined inside https://github.com/GoogleCloudPlatform/ops-agent/tree/master/confgenerator

- One file per application integration, defining receiver and processor struct(s) as needed for that application including both logging and metrics
- Validation is performed with struct tags on each config struct (~zero per-type validation code)
- Config structs have a method to generate OT and fluentbit config from the config struct. See details in the Pipelines() method below

> - Each application register its own receiver by specifying the receiver in the init() function of its own file. Adding support for a new application only needs to touch this one new file, unless it needs to add common utils.

Reviewer: ```suggestion
- Each application registers its own receiver by specifying the receiver in the init() function of its own file. Adding support for a new application only needs to touch this one new file, unless it needs to add common utils.
```

Example code for IIS application

```go
package confgenerator

import "github.com/GoogleCloudPlatform/ops-agent/confgenerator/otel"

type MetricsReceiverIis struct {
	ConfigComponent `yaml:",inline"`

> 	MetricsReceiverShared `yaml:",inline"`

Reviewer 2: ```suggestion
	// This is a convenience struct with common fields like CollectionInterval.
	MetricsReceiverShared `yaml:",inline"`
```

}

func (r MetricsReceiverIis) Type() string {

> 	return "iis"

Reviewer 2: ```suggestion
	// This is the string that will identify this receiver in user configs.
	return "iis"
```
